### PR TITLE
Allow `!ban`, `!vban`, and `!pban` to take a duration

### LINKED ISSUE #1563: When a duration is specified for !ban, it should become a tempban
# Proposal
When a duration is specified for !ban like: `!ban <id> 3D You are now banned.` It will become a temporary ban for the specified duration.
`!ban` with no provided duration: `!ban <id> You are now banned.` should still result in a permanent ban.

# Reasoning
Myself and other moderators will sometimes be typing out the ban when we realize we want it to be a tempban. We then need to backtrack and change it to tempban, which is a nuisance.

Additionally, the `!mute` command works with a duration, turning it into a tempmute. (In fact the default argument is a 1H mute.) It would be nice if the two could behavior similarly where if a duration is specified, it becomes temporary.

### LINKED ISSUE #1566: Pban will use duration for tempbanning, not purge_days. Also tpban should be a thing.
# Proposal
- Currently `!pban` can be provided a duration to specify how many days of messages to purge (ranging from 1 to 7) with the default being 1 day. I don't think we have every actually used a duration other than 1 day, which is the default.
- In the spirit of consistency, we would like the option to pass in a duration for allow tempbanning of the person in addition to purging messages.
- So we want to remove the purge_days argument (always make it purge 1 days work of messages) and if a duration is specified, it tempbans the person for that duration
- The above should also exist as its own standalone command: `!tpban`

# Reasoning
- Consistency! We're altering how `!ban` and `!vban` work to allow for optional arguments for duration that would turn them into temporary bans. It would be nice if all of our ban commands operated similarly to reduce mod command mishaps.
- While we do strive to Delete + Report the first message of something that breaks discord's rules, it isn't really feasible to manually delete a large quantity of messages. After we delete + report the first message, this command will essentially allow us to clean + tempban / clean + ban in one go which is nice.

### PR DESCRIPTION
**Pull request by Shivansh-007**
Closes #1563 
Closes #1566 

Apply temporary ban if duration is specified while banning a user.

#### Applying a `ban` with duration
Screenshot from 2021-05-05 06-45-34

#### Proof in `#all-logs`
Screenshot from 2021-05-05 06-46-26

### CODE REVIEW

File: bot/exts/moderation/infraction/infractions.py [hunk 1]
```diff
@@ -76,9 +76,19 @@ async def kick(self, ctx: Context, user: FetchedMember, *, reason: t.Optional[st
...
+        duration: t.Optional[Expiry] = None,
+        *,
+        reason: t.Optional[str] = None
+    ) -> None:
         """Permanently ban a user for the given reason and stop watching them with Big Brother."""
-        await self.apply_ban(ctx, user, reason)
+        if duration:
+            await self.apply_ban(ctx, user, reason, expires_at=duration)
+        else:
+            await self.apply_ban(ctx, user, reason)
```
Comment (ToxicKidz) [inline bot/exts/moderation/infraction/infractions.py:91]:
```suggestion
        await self.apply_ban(ctx, user, reason, expires_at=duration)
```

**Review by ToxicKidz — CHANGES_REQUESTED** [review]

File: bot/exts/moderation/infraction/infractions.py [hunk 2]
```diff
@@ -76,9 +76,16 @@ async def kick(self, ctx: Context, user: FetchedMember, *, reason: t.Optional[st
...
-    async def ban(self, ctx: Context, user: FetchedMember, *, reason: t.Optional[str] = None) -> None:
+    async def ban(
+        self,
+        ctx: Context,
+        user: FetchedMember,
+        duration: t.Optional[Expiry] = None,
+        *,
+        reason: t.Optional[str] = None
+    ) -> None:
         """Permanently ban a user for the given reason and stop watching them with Big Brother."""
```
Comment (Xithrius) [inline bot/exts/moderation/infraction/infractions.py:87]: Docstring should be updated to correlate with this new feature. Using the ban command will not always end up as a permanent ban.

Comment (Shivansh-007) [reply bot/exts/moderation/infraction/infractions.py]: Fixed in commit ea900a4.

**Review by Xithrius — APPROVED** [review]

**Review by ToxicKidz — APPROVED** [review]
:shipit:

**Review by mbaruh — APPROVED** [review]

**Comment by mbaruh** [comment]
It looks good in the context of the issue, though it creates inconsistency with pban and vban

**Review by Xithrius — CHANGES_REQUESTED** [review]
To make this consistent with other commands, please make purgeban not take any argument for purge days, but instead take only ban days. If time exists, don't permaban.

Please also do the same thing for voiceban, where if a time is supplied, it won't be permanent.

**Comment by mbaruh** [comment]
You're still going over the character limit for commit titles :)
Keep them concise. Any details of context or reasoning can go into the body.

File: bot/exts/moderation/infraction/infractions.py [hunk 3]
```diff
@@ -76,31 +76,52 @@ async def kick(self, ctx: Context, user: FetchedMember, *, reason: t.Optional[st
...
         ctx: Context,
         user: FetchedMember,
-        purge_days: t.Optional[int] = 1,
+        duration: t.Optional[Expiry] = None,
         *,
         reason: t.Optional[str] = None
     ) -> None:
         """
-        Same as ban but removes all their messages for the given number of days, default being 1.
+        Same as ban but removes all their messages of the current day.
```
Comment (mbaruh) [inline bot/exts/moderation/infraction/infractions.py:104]:
```suggestion
        Same as ban but removes all their messages of the last 24 hours.
```
"Current day" can be misleading.

File: bot/exts/moderation/infraction/infractions.py [hunk 4]
```diff
@@ -76,31 +76,52 @@ async def kick(self, ctx: Context, user: FetchedMember, *, reason: t.Optional[st
...
+        ctx: Context,
+        user: FetchedMember,
+        duration: t.Optional[Expiry] = None,
+        *,
+        reason: t.Optional[str] = None
+    ) -> None:
+        """
+        Permanently ban a user for the given reason and stop watching them with Big Brother.
+
+        If duration is specified, then it would temporarily ban that user for the given duration.
```
Comment (mbaruh) [inline bot/exts/moderation/infraction/infractions.py:90]:
```suggestion
        If duration is specified, it temporarily bans that user for the given duration.
```

File: bot/exts/moderation/infraction/infractions.py [hunk 5]
```diff
@@ -76,31 +76,52 @@ async def kick(self, ctx: Context, user: FetchedMember, *, reason: t.Optional[st
...
+        ctx: Context,
+        user: FetchedMember,
+        duration: t.Optional[Expiry] = None,
+        *,
+        reason: t.Optional[str]
+    ) -> None:
+        """
+        Permanently ban user from using voice channels.
+
+        If duration is specified, then it would temporarily voice ban that user for the given duration.
```
Comment (mbaruh) [inline bot/exts/moderation/infraction/infractions.py:122]:
```suggestion
        If duration is specified, it temporarily voice bans that user for the given duration.
```

**Review by mbaruh — CHANGES_REQUESTED** [review]

File: bot/exts/moderation/infraction/infractions.py [hunk 6]
```diff
@@ -76,31 +76,52 @@ async def kick(self, ctx: Context, user: FetchedMember, *, reason: t.Optional[st
...
         """
-        Same as ban but removes all their messages for the given number of days, default being 1.
+        Same as ban but removes all their messages of the last 24 hours.
 
-        `purge_days` can only be values between 0 and 7.
-        Anything outside these bounds are automatically adjusted to their respective limits.
+        If duration is specified, it temporarily bans that user for the given duration.
         """
-        await self.apply_ban(ctx, user, reason, max(min(purge_days, 7), 0))
+        await self.apply_ban(ctx, user, reason, 1, 0, expires_at=duration)
```
Comment (mbaruh) [inline bot/exts/moderation/infraction/infractions.py:108]: What is that 0 for? please test changes before pushing them ^^"

**Review by mbaruh — CHANGES_REQUESTED** [review]

File: bot/exts/moderation/infraction/infractions.py (hunk 6 above)
Comment (Shivansh-007) [reply bot/exts/moderation/infraction/infractions.py]: Did test the others, missed out on this one.

**Review by Xithrius — APPROVED** [review]
Looks very well to me.

**Comment by Shivansh-007** [comment]
> This can be an opportunity to add some more related tests, but you don't have to.

Tests for the new _feature_ I added in this PR, or the complete `pban` and `ban` command?

**Comment by mbaruh** [comment]
For what you added.

**Comment by Shivansh-007** [comment]
> For what you added.

Alright, I will do that.

**Pull request merged**